builder/llpcBuilderImplSubgroup.cpp

```diff
@@ -0,0 +1,1185 @@
+                                getInt32(bankMask),
+                                getInt1(boundCtrl)
+                          });
+}
+
+// =====================================================================================================================
+// Create a call to dpp update.
+Value* BuilderImplSubgroup::CreateDPPUpdate(
+    Value* const pOriginal, // [in] The original value we are going to update.
+    Value* const pValue,    // [in] The value to DPP update.
+    DPPCtrl      dppCtrl,   // The dpp_ctrl to use.
+    unsigned     rowMask,   // The row mask.
+    unsigned     bankMask,  // The bank mask.
+    bool         boundCtrl) // Whether bound_ctrl is used or not.
+{
+    auto function = [](Builder& builder, ArrayRef<Value*> mappedArgs, ArrayRef<Value*> passthroughArgs) -> Value*
+    {
+        return builder.CreateIntrinsic(Intrinsic::amdgcn_update_dpp,
+                                       builder.getInt32Ty(),
+                                       {
+                                            mappedArgs[0],
+                                            mappedArgs[1],
+                                            passthroughArgs[0],
+                                            passthroughArgs[1],
+                                            passthroughArgs[2],
+                                            passthroughArgs[3]
+                                       });
+    };
+
+    return CreateMapToI32(function,
+                          {
+                                pOriginal,
+                                pValue,
+                          },
+                          {
+                                getInt32(static_cast<uint32_t>(dppCtrl)),
+                                getInt32(rowMask),
+                                getInt32(bankMask),
+                                getInt1(boundCtrl)
+                          });
+}
+
+// =====================================================================================================================
+// Create a call to ds swizzle.
+Value* BuilderImplSubgroup::CreateDSSwizzle(
+    Value* const pValue, // [in] The value to swizzle.
+    uint16_t dsPattern)  // The pattern to swizzle with.
+{
+    auto function = [](Builder& builder, ArrayRef<Value*> mappedArgs, ArrayRef<Value*> passthroughArgs) -> Value*
+    {
+        return builder.CreateIntrinsic(Intrinsic::amdgcn_ds_swizzle,
+                                       {},
+                                       {
+                                            mappedArgs[0],
+                                            passthroughArgs[0]
+                                       });
+    };
+
+    return CreateMapToI32(function, pValue, getInt32(dsPattern));
+}
+
+// =====================================================================================================================
+// Create a call to WWM (whole wave mode).
+Value* BuilderImplSubgroup::CreateWWM(
```

Comment:
Can we use CreateMapToI32 for CreateWWM instead of do recursive by itself?

Reply:
Yes - I don't know why I didn't and I've added that!